### DISCUSSION TITLE
Adding a tutorial/guide for periodic backups

### DISCUSSION THREAD

**Discussion by ClementNerma** [firefly-iii/firefly-iii#4310] [Firefly III]
As Firefly III stores very important informations that most of us can't afford to lose, I think it'd be great to add a little guide or tutorial to indicate how to backup the whole data.

Something that comes to my mind, which is my actual backup process, it to complete the reference `docker-compose.yml` with a backup container, like [https://github.com/fradelg/docker-mysql-cron-backup](this one) for instance.

Here is my personal Docker Compose file for reference:

```yaml
version: "3.3"

services:
  fireflyiii:
    container_name: software_fireflyiii
    image: jc5x/firefly-iii:latest
    volumes:
      - firefly_iii_upload:/var/www/html/storage/upload
    env_file: .env
    ports:
      - 7000:8080
    depends_on:
      - fireflyiiidb

  fireflyiiidb:
    container_name: software_fireflyiii_db
    image: yobasystems/alpine-mariadb:latest
    environment:
      - MYSQL_RANDOM_ROOT_PASSWORD=yes
      - MYSQL_USER=XXXXXXXXXX
      - MYSQL_PASSWORD=XXXXXXXXXX
      - MYSQL_DATABASE=XXXXXXXXXX
    volumes:
      - firefly_iii_db:/var/lib/mysql

  mysql-cron-backup:
    container_name: software_fireflyiii_db_backuper
    image: fradelg/mysql-cron-backup
    depends_on:
      - fireflyiiidb
    volumes:
      - ./mysql_backup:/backup
    environment:
      - MYSQL_HOST=software_fireflyiii_db
      - MYSQL_USER=XXXXXXXXXX
      - MYSQL_PASS=XXXXXXXXXX
      - MAX_BACKUPS=15
      - INIT_BACKUP=1
      - CRON_TIME=0 * * * * # Every hour at minute 0
      - GZIP_LEVEL=9 # High compression level

volumes:
  firefly_iii_upload:
  firefly_iii_db:
```

Of course the period can be changed, but here the database is completely backed up every hour. With the current database I have (not so many transactions) it takes around 25 KB compressed. The backup consists in a simple SQL file that can be run to restore the database to the state it was at when the backup was performed.

What do you think of it?

**Comment by JC5** [comment]
Generally speaking, I want to leave this to people to handle themselves. There are many different solutions to backup (container) data and I can't add stuff to the documentation that people can Google as well. Having said that, I don't mind adding some more hints and pointers to [this page](https://docs.firefly-iii.org/firefly-iii/advanced-installation/backup/).

**Comment by ClementNerma** [comment]
> Generally speaking, I want to leave this to people to handle themselves. There are many different solutions to backup (container) data and I can't add stuff to the documentation that people can Google as well.

I've read that on several issues already, and I don't agree with that point of view ; it should straightforward and easy to backup and restore a database containing such important informations. Having to look for a solution and experiment, especially when we're afraid of messing up by trying some things when the database already contains a lot of data, is only a frustration factor.

I think there should even be a simple full backup / restore option in the app, but just showing a simple backup system like this can be a good thing (it took me some time before finding it when I was looking for a simple backup solution).

> I don't mind adding some more hints and pointers to [this page](https://docs.firefly-iii.org/firefly-iii/advanced-installation/backup/).

That's exactly what I meant ^^

**Comment by lucmos** [comment]
Hey, just stumbled here in search of backup best practices!
I have a docker instance of Firefly III. I'm not familiar with docker.

Following the docs:
> If you're running Firefly III in Docker, backup the following:
> 
> - [x] The Docker variables you've used to launch the container (s), and especially the APP_KEY-variable.
> - [x] Any secrets you have used (database passwords).
> - [ ] The two volumes used by Firefly III: "upload" and "db"

I have some doubts about the third point. At the moment I have a crontab restic entry to backup the entire server with firefly's volumes on it, every day.

Is it safe to backup live volumes? I did read that's suggested to do filesystem snapshots and backup them, to avoid inconsistent and corrupted states/databases. Is that overkill? I'd like to avoid this for my old homeserver.


For my needs, it would be enough a periodic local backup of firefly to enforce consistency. This local backup will eventually be backup-ed on cloud by restic (I am currently doing this with a nextcloud instance, that offers an "autobackup" option from the control panel, and it seems to work fine).
Is it possible to do this with firefly?


How can I safely backup the volumes of a running instance of firefly periodically?

Is the command suggested on the docs safe to run on live volumes?
`docker run --rm -v "firefly_iii_db:/tmp" -v "$HOME/backups/firefly:/backup" ubuntu tar -czvf /backup/firefly_db.tar /tmp`

**Comment by ClementNerma** [comment]
I don't know about how volumes backup exactly work, but the safest way is to simply backup the volume's _content_ instead of the volume itself. Which is what the command in the docs does.

**Comment by lucmos** [comment]
Ok great!

I mount the volume on the file system (on a raid), so it is equivalent in that regard. i.e. in the docker-compose:
```
volumes:
      - /appdata/fireflyiii/db:/var/lib/mysql
...
```


My doubt was about performing the backup *while* firefly is running. Is it okay?

Since the state of the files/db captured may be more or less probable to be inconsistent, depending on how firefly works.

**Comment by niklas2810** [comment]
My take on this is to perform backups when I don't use the instance (e.g. at 4AM). I copy the entire folder, compress it into a .tar.gz file and encrypt it using a PGP key (and backup this file to an external storage device).

I can't actually say whether backing it up in an inactive state is failure-proof, but at least I can confirm that my backups were still usable whenever I used them.

**Comment by ClementNerma** [comment]
> Ok great!
> 
> I mount the volume on the file system (on a raid), so it is equivalent in that regard. i.e. in the docker-compose:
> 
> ```
> volumes:
>       - /appdata/fireflyiii/db:/var/lib/mysql
> ...
> ```
> 
> My doubt was about performing the backup _while_ firefly is running. Is it okay?
> 
> Since the state of the files/db captured may be more or less probable to be inconsistent, depending on how firefly works.

Unless you have Cron tasks running, Firefly won't be running in the background so no worries.

**Reply by devfaz** [reply]
No worries is a bit too easy. It's quite unlikely, but you may still run into inconsistent database files. The only real safe way is doing an dump and/or do some kind of "write freeze" before doing the copy of the binary database files.

**Reply by ClementNerma** [reply]
No because the database isn't called at this time, and databases (to my knowledge at least) don't perform background actions on their own.
So there's no problem with this.
But yes a dump is always better, which the solution I show in my first comment ;)

**Reply by lucmos** [reply]
Eventually, I opted to use a BTRFS filesystem. My backup routine creates a snapshot of the volume, and only that (freezed) snapshot is under automatic backup.

**Reply by devfaz** [reply]
> No because the database isn't called at this time, and databases (to my knowledge at least) don't perform background actions on their own.
> So there's no problem with this.
> But yes a dump is always better, which the solution I show in my first comment ;)

--

> Eventually, I opted to use a BTRFS filesystem. My backup routine creates a snapshot of the volume, and only that (freezed) snapshot is under automatic backup.

--

both will work 99,999% of the time, but its the 0,001% you need when your data is gone ;)

Backup is easy - restore is what matters!

**Comment by charly127** [comment]
Hello

What would the reverse process look like if the system crashes and all data is deleted? 

Thank you

**Reply by ClementNerma** [reply]
Check out https://github.com/fradelg/docker-mysql-cron-backup

Basically:

```shell
# List available backups
docker container exec <your_mysql_backup_container_name> ls /backup

# Restore a specific backup
docker container exec <your_mysql_backup_container_name> /restore.sh /backup/<your_sql_backup_gz_file>
```

**Comment by jonashackt** [comment]
I just lately stumbled upon firefly III and I think it's an amazing tooling, so I set it up for me and my wife on a homeserver. The issue on how to backup came soon and [I read the docs](https://docs.firefly-iii.org/how-to/firefly-iii/advanced/backup/#automated-backup-using-a-bash-script-and-crontab), where the great script by [@dawid-czarnecki](https://github.com/dawid-czarnecki) is linked. This script works well for me to do a local backup, but I wanted to have the backup been synced to another server outside my house to preserve data in case of hardware defects (such as Google Drive).

On my way I learned, that cron has kind of a defacto successor in modern distros: when I looked into the Arch Wiki I found that on Arch (and Manjaro, which I use) [the cron service isn't even enabled per default anymore, "as the base system uses systemd/Timers instead"](https://wiki.archlinux.org/title/Cron). So I thought about switching to Systemd instead and it was a great decisision for me, since I got all the benefits of a modern system integrated service scheduling with great logging and monitoring freatures (see [this list of benefits of using systemd Timers instead of cron in the Arch Wiki](https://wiki.archlinux.org/title/Systemd/Timers#As_a_cron_replacement) also).

**TLDR: If anyone is interested, I put together [a guide on how to use @dawid-czarnecki's backup script together with openssl (encryption), rclone (Google Drive sync) and systemd Timers (scheduled execution of backup)](https://github.com/jonashackt/firefly-iii-ops/blob/main/README.md#backup-firefly-iii).**

Just as a quick glance, here's my `fireflybackup.service` file in `/etc/systemd/system`:

```shell
[Unit]
Description=Service for Firefly III Backup to Google Drive
Wants=fireflybackup.timer

[Service]
Type=oneshot
ExecStart=/home/homepike/firefly/fireflybackup.sh
User=homepike
Group=homepike

[Install]
WantedBy=timers.target
```

And here's my `fireflybackup.timer` also in `/etc/systemd/system/` running my firefly backup every day at 10:00am and 5min after boot:

```shell
[Unit]
Description=Timer unit for fireflybackup.service
Requires=fireflybackup.service

[Timer]
Unit=fireflybackup.service
OnCalendar=*-*-* 10:00:00
OnBootSec=5min

[Install]
WantedBy=timers.target
```
